File: packages/core/src/ops.ts
/**
 * Pure branching operations shared by every Deku backend (Express, browser, and
 * — via the JSON fixtures in ../fixtures/ops — the Python port).
 * No IO, no environment: everything operates on plain Conversation objects.
 */
import { estimateTokens, messageTokens } from './types.ts';
import type { Branch, Conversation, ConversationSummary, Msg, Role } from './types.ts';

const uuid = () => globalThis.crypto.randomUUID();

export function makeMsg(role: Role, content: string, id: string = uuid()): Msg {
  return { id, role, kind: 'message', content, createdAt: Date.now() };
}

export function newConversation(id: string = uuid()): Conversation {
  const now = Date.now();
  const main: Branch = { id: uuid(), name: 'Main', forkOf: null, messages: [], createdAt: now };
  return { id, title: 'New chat', branches: [main], activeBranchId: main.id, createdAt: now, updatedAt: now };
}

export function touch(conv: Conversation): void {
  conv.updatedAt = Date.now();
}

export function getBranch(conv: Conversation, branchId: string): Branch | undefined {
  return conv.branches.find((b) => b.id === branchId);
}

/**
 * New branch containing `from`'s messages up to and including messageId.
 * messageId === null means "full copy" (plain Branch); forkOf.messageId is then the tip.
 * Mutates conv (pushes the branch) and returns it.
 */
export function forkBranch(conv: Conversation, from: Branch, messageId: string | null, name: string): Branch {
  let prefix = from.messages;
  if (messageId !== null) {
    const idx = from.messages.findIndex((m) => m.id === messageId);
    if (idx === -1) throw new Error('message not found in branch');
    prefix = from.messages.slice(0, idx + 1);
  }
  const branch: Branch = {
    id: uuid(),
    name,
    forkOf: { branchId: from.id, messageId: messageId ?? from.messages.at(-1)?.id ?? null },
    messages: prefix.map((m) => ({ ...m })),
    createdAt: Date.now(),
  };
  conv.branches.push(branch);
  return branch;
}

export function nextName(conv: Conversation, base: 'Branch' | 'Rewind'): string {
  const n = conv.branches.filter((b) => b.name.startsWith(base)).length + 1;
  return `${base} ${n}`;
}

/**
 * Replay `source` onto `target`: append source's messages target lacks (by id),
 * re-point source's children at target, delete source, activate target.
 * Throws on unknown/identical branches.
 */
export function combineBranches(conv: Conversation, sourceBranchId: string, targetBranchId: string): void {
  const source = getBranch(conv, sourceBranchId);
  const target = getBranch(conv, targetBranchId);
  if (!source || !target || source.id === target.id) throw new Error('bad branches');
  const targetIds = new Set(target.messages.map((m) => m.id));
  target.messages.push(...source.messages.filter((m) => !targetIds.has(m.id)).map((m) => ({ ...m })));
  for (const b of conv.branches) {
    if (b.forkOf?.branchId === source.id) b.forkOf.branchId = target.id;
  }
  conv.branches = conv.branches.filter((b) => b.id !== source.id);
  conv.activeBranchId = target.id;
}

/**
 * Replace branch messages [fromId..toId] with a single summary node carrying the
 * originals and the freed-token estimate (floored at 0). Returns the summary node.
 * Throws 'bad range' on unknown ids or inverted range.
 */
export function spliceSummary(branch: Branch, fromId: string, toId: string, summaryText: string): Msg {
  const lo = branch.messages.findIndex((m) => m.id === fromId);
  const hi = branch.messages.findIndex((m) => m.id === toId);
  if (lo === -1 || hi === -1 || lo > hi) throw new Error('bad range');
  const originals = branch.messages.slice(lo, hi + 1);
  const freed = originals.reduce((n, m) => n + messageTokens(m), 0) - estimateTokens(summaryText);
  const summary: Msg = {
    ...makeMsg('assistant', summaryText),
    kind: 'summary',
    summaryOf: originals,
    freedTokens: Math.max(freed, 0),
  };
  branch.messages.splice(lo, originals.length, summary);
  return summary;
}

/** last-resort summary when no LLM is reachable */
export function summaryFallback(msgs: Msg[]): string {
  const firstLine = (s: string) => s.split('\n')[0].slice(0, 90);
  return `${firstLine(msgs[0].content)} → ${firstLine(msgs.at(-1)!.content)}`;
}

export function toSummary(c: Conversation): ConversationSummary {
  return { id: c.id, title: c.title, branchCount: c.branches.length, updatedAt: c.updatedAt };
}

File: apps/fullstack/backend/deku_server/ops.py
"""Port of packages/core/src/ops.ts — kept in lockstep by the shared JSON fixtures
in packages/core/fixtures/ops (see tests/test_ops.py)."""
import math
import time
import uuid

from .models import Branch, Conversation, ConversationSummary, ForkOf, Msg


def now_ms() -> int:
    return int(time.time() * 1000)


def _uuid() -> str:
    return str(uuid.uuid4())


def estimate_tokens(text: str) -> int:
    """rough universal token estimate: ~4 chars per token"""
    return math.ceil(len(text) / 4)


def message_tokens(m: Msg) -> int:
    tools = sum(estimate_tokens(t.label + t.detail) for t in (m.tool_events or []))
    return estimate_tokens(m.content) + tools + 4


def make_msg(role: str, content: str, id: str | None = None) -> Msg:
    return Msg(id=id or _uuid(), role=role, kind="message", content=content, created_at=now_ms())


def new_conversation(id: str | None = None) -> Conversation:
    now = now_ms()
    main = Branch(id=_uuid(), name="Main", fork_of=None, messages=[], created_at=now)
    return Conversation(
        id=id or _uuid(), title="New chat", branches=[main],
        active_branch_id=main.id, created_at=now, updated_at=now,
    )


def touch(conv: Conversation) -> None:
    conv.updated_at = now_ms()


def get_branch(conv: Conversation, branch_id: str) -> Branch | None:
    return next((b for b in conv.branches if b.id == branch_id), None)


def fork_branch(conv: Conversation, from_branch: Branch, message_id: str | None, name: str) -> Branch:
    """New branch with from_branch's messages up to and including message_id.
    message_id None = full copy (plain Branch); forkOf.messageId is then the tip."""
    prefix = from_branch.messages
    if message_id is not None:
        idx = next((i for i, m in enumerate(from_branch.messages) if m.id == message_id), -1)
        if idx == -1:
            raise ValueError("message not found in branch")
        prefix = from_branch.messages[: idx + 1]
    tip = from_branch.messages[-1].id if from_branch.messages else None
    branch = Branch(
        id=_uuid(),
        name=name,
        fork_of=ForkOf(branch_id=from_branch.id, message_id=message_id if message_id is not None else tip),
        messages=[m.model_copy() for m in prefix],
        created_at=now_ms(),
    )
    conv.branches.append(branch)
    return branch


def next_name(conv: Conversation, base: str) -> str:
    n = len([b for b in conv.branches if b.name.startswith(base)]) + 1
    return f"{base} {n}"


def combine_branches(conv: Conversation, source_branch_id: str, target_branch_id: str) -> None:
    """Append source's messages target lacks (by id), re-point source's children
    at target, delete source, activate target."""
    source = get_branch(conv, source_branch_id)
    target = get_branch(conv, target_branch_id)
    if source is None or target is None or source.id == target.id:
        raise ValueError("bad branches")
    target_ids = {m.id for m in target.messages}
    target.messages.extend(m.model_copy() for m in source.messages if m.id not in target_ids)
    for b in conv.branches:
        if b.fork_of is not None and b.fork_of.branch_id == source.id:
            b.fork_of.branch_id = target.id
    conv.branches = [b for b in conv.branches if b.id != source.id]
    conv.active_branch_id = target.id


def splice_summary(branch: Branch, from_id: str, to_id: str, summary_text: str) -> Msg:
    """Replace [from_id..to_id] with a summary node keeping the originals and the
    freed-token estimate (floored at 0)."""
    lo = next((i for i, m in enumerate(branch.messages) if m.id == from_id), -1)
    hi = next((i for i, m in enumerate(branch.messages) if m.id == to_id), -1)
    if lo == -1 or hi == -1 or lo > hi:
        raise ValueError("bad range")
    originals = branch.messages[lo : hi + 1]
    freed = sum(message_tokens(m) for m in originals) - estimate_tokens(summary_text)
    base = make_msg("assistant", summary_text)
    summary = base.model_copy(
        update={"kind": "summary", "summary_of": originals, "freed_tokens": max(freed, 0)}
    )
    branch.messages[lo : hi + 1] = [summary]
    return summary


def summary_fallback(msgs: list[Msg]) -> str:
    def first_line(s: str) -> str:
        return s.split("\n")[0][:90]

    return f"{first_line(msgs[0].content)} → {first_line(msgs[-1].content)}"


def to_summary(c: Conversation) -> ConversationSummary:
    return ConversationSummary(
        id=c.id, title=c.title, branch_count=len(c.branches), updated_at=c.updated_at
    )
